authority/provisioner/webhook.go

```diff
@@ -0,0 +1,203 @@
+}
+
+type WebhookResponseBody struct {
+	Data  map[string]interface{} `json:"data"`
+	Allow bool                   `json:"allow"`
+}
+
+func (w *Webhook) Do(ctx context.Context, client *http.Client, certReq interface{}, data map[string]interface{}) (*WebhookResponseBody, error) {
+	tmpl, err := template.New("url").Funcs(templates.StepFuncMap()).Parse(w.URL)
+	if err != nil {
+		return nil, err
+	}
+	buf := &bytes.Buffer{}
+	if err := tmpl.Execute(buf, data); err != nil {
+		return nil, err
+	}
+	url := buf.String()
+
+	ctx, cancel := context.WithTimeout(ctx, time.Second*10)
+	defer cancel()
+	retries := 1
+retry:
+	reqBody := &webhookRequestBody{
+		Timestamp: time.Now(),
+	}
+	switch r := certReq.(type) {
+	case *x509.CertificateRequest:
+		if r != nil {
+			reqBody.X509_CSR = r.Raw
+		}
+	case sshutil.CertificateRequest:
+		reqBody.SSH_CR = &sshRequest{
+			Type:       r.Type,
+			KeyID:      r.KeyID,
+			Principals: r.Principals,
+		}
+		if r.Key != nil {
+			reqBody.SSH_CR.Key = r.Key.Marshal()
+		}
+	case *x509.Certificate:
+		reqBody.Certificate = x509util.FromX509Certificate(r)
+	case *ssh.Certificate:
+		reqBody.SSHCertificate = &sshCert{
+			Nonce:           r.Nonce,
+			Key:             r.Key.Marshal(),
+			Serial:          r.Serial,
+			CertType:        r.CertType,
+			KeyID:           r.KeyId,
+			ValidPrincipals: r.ValidPrincipals,
+			ValidAfter:      r.ValidAfter,
+			ValidBefore:     r.ValidBefore,
+			Permissions:     r.Permissions,
+			Reserved:        r.Reserved,
+		}
+		if r.Key != nil {
+			reqBody.SSHCertificate.Key = r.Key.Marshal()
+		}
+		if r.SignatureKey != nil {
+			reqBody.SSHCertificate.SignatureKey = r.SignatureKey.Marshal()
+		}
+	}
+
+	reqBytes, err := json.Marshal(reqBody)
+	if err != nil {
+		return nil, err
+	}
+
+	req, err := http.NewRequestWithContext(ctx, "POST", url, bytes.NewReader(reqBytes))
+	if err != nil {
+		return nil, err
+	}
+
+	secret, err := base64.StdEncoding.DecodeString(w.SigningSecret)
+	if err != nil {
+		return nil, err
+	}
+	sig := hmac.New(sha256.New, secret).Sum(reqBytes)
+	req.Header.Set("X-Smallstep-Signature", hex.EncodeToString(sig))
+	req.Header.Set("X-Smallstep-Webhook-ID", w.ID)
+
+	if w.BearerToken != "" {
+		req.Header.Set("Authorization", fmt.Sprintf("Bearer %s", w.BearerToken))
+	} else if w.BasicAuth.Username != "" || w.BasicAuth.Password != "" {
+		req.SetBasicAuth(w.BasicAuth.Username, w.BasicAuth.Password)
+	}
+
+	if w.DisableTLSClientAuth {
+		transport, ok := client.Transport.(*http.Transport)
+		if !ok {
+			return nil, errors.New("client transport is not a *http.Transport")
+		}
+		transport = transport.Clone()
+		tlsConfig := transport.TLSClientConfig.Clone()
+		tlsConfig.GetClientCertificate = nil
+		tlsConfig.Certificates = nil
+		transport.TLSClientConfig = tlsConfig
+		client = &http.Client{
+			Transport: transport,
+		}
+	}
+	resp, err := client.Do(req)
+	if err != nil {
+		if err == context.DeadlineExceeded {
```

Comment:
The linter might complain unless you do `errors.Is(err, context.DeadlineExceeded)`

Reply:
f6205018d8bcc271b71f82dd37ceef17c001f84d